processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,481 @@
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      while ((nextWriteIndex == nextReadIndex) && !Thread.currentThread().isInterrupted()) {
+        if (now >= queryTimeoutAtNs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+        // Thread.yield() should not be called from the very beginning
+        if (now >= spinTimeoutAt) {
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = nextWriteIndex;
+    }
+  }
+
+  @Override
+  public void reset()
+  {
+    curWriteIndex = -1;
+    nextReadIndex = -1;
+    finished = false;
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  /**
+   * Signal that no more inputs are added.  Must be called after {@link #aggregate(Object)} is called for the last input.
+   */
+  public void finish()
+  {
+    increaseWriteIndex();
+    finished = true;
+  }
+
+  /**
+   * Return a sorted iterator.  This method can be called safely while writing, and the iterating thread and the writing
+   * thread can be different.  The result iterator always returns sorted results.  This method should be called only one
+   * time per grouper.
+   *
+   * @return a sorted iterator
+   */
+  public CloseableIterator<Entry<KeyType>> iterator()
+  {
+    if (!initialized) {
+      throw new ISE("Grouper should be initialized first");
+    }
+
+    return new CloseableIterator<Entry<KeyType>>()
+    {
+      {
+        // Wait for some data to be ready and initialize nextReadIndex.
+        increaseReadIndexTo(0);
+      }
+
+      @Override
+      public boolean hasNext()
+      {
+        // Once finished becomes true, curWriteIndex isn't changed anymore and thus remainig() can be computed safely
+        // because nextReadIndex is changed only by the reading thread.
+        return !finished || remaining() > 0;
+      }
+
+      /**
+       * Calculate the number of remaining items in the array.  Must be called only when
+       * {@link StreamingMergeSortedGrouper#finished} is true.
+       *
+       * @return the number of remaining items
+       */
+      private int remaining()
+      {
+        if (curWriteIndex >= nextReadIndex) {
+          return curWriteIndex - nextReadIndex;
+        } else {
+          return (maxNumSlots - nextReadIndex) + curWriteIndex;
+        }
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        final int recordOffset = recordSize * nextReadIndex;
+        final KeyType key = keySerde.fromByteBuffer(buffer, recordOffset);
```

Comment:
I don't see HB between the writing thread and the reading thread, the reading thread may not avoid reading `curWriteIndex` as far as I can see, and then the visibility of the buffer contents is not guaranteed.

Reply:
Changes on curWriteIndex happens-before changing nextReadIndex. Added more comments.